sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3Path.java

```diff
@@ -0,0 +1,569 @@
+import java.nio.file.FileSystem;
+import java.nio.file.LinkOption;
+import java.nio.file.Path;
+import java.nio.file.WatchEvent.Kind;
+import java.nio.file.WatchEvent.Modifier;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Iterator;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.coders.SerializableCoder;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+/**
+ * Implements the Java NIO {@link Path} API for AWS S3 paths.
+ */
+public class S3Path implements Path, Serializable {
+
+  public static final String SCHEME = "s3";
+
+  /**
+   * Matches a glob containing a wildcard, capturing the portion before the first wildcard.
+   */
+  private static final Pattern GLOB_PREFIX = Pattern.compile("(?<PREFIX>[^\\[*?]*)[\\[*?].*");
+
+  public static S3Path fromUri(URI uri) {
+    checkArgument(uri.getScheme().equals(SCHEME), "URI: %s is not an S3 URI", uri);
+    checkArgument(uri.getPort() == -1, "S3 URI may not specify port: %s (%i)", uri, uri.getPort());
+    checkArgument(
+        Strings.isNullOrEmpty(uri.getUserInfo()),
+        "S3 URI may not specify userInfo: %s (%s)",
+        uri,
+        uri.getUserInfo());
+    checkArgument(
+        Strings.isNullOrEmpty(uri.getQuery()),
+        "S3 URI may not specify query: %s (%s)",
+        uri,
+        uri.getQuery());
+    checkArgument(
+        Strings.isNullOrEmpty(uri.getFragment()),
+        "S3 URI may not specify fragment: %s (%s)",
+        uri,
+        uri.getFragment());
+
+    String bucket = uri.getHost();
+    String key = uri.getPath();
+    if (key.startsWith("/")) {
+      key = key.substring(1);
+    }
+    return fromComponents(bucket, key);
+  }
+
+  private static final Pattern S3_URI =
+      Pattern.compile("(?<SCHEME>[^:]+)://(?<BUCKET>[^/]+)(/(?<OBJECT>.*))?");
+
+  public static S3Path fromUri(String uri) {
+    Matcher m = S3_URI.matcher(uri);
+    checkArgument(m.matches(), "Invalid S3 URI: %s", uri);
+
+    checkArgument(m.group("SCHEME").equalsIgnoreCase(SCHEME), "URI: %s is not an S3 URI", uri);
+    return new S3Path(null, m.group("BUCKET"), m.group("OBJECT"));
+  }
+
+  public static S3Path fromComponents(String bucket, String key) {
+    return new S3Path(null, bucket, key);
+  }
+
+  @Nullable
+  private transient FileSystem fileSystem;
+  @Nonnull
+  private final String bucket;
+  @Nonnull
+  private final String key;
+
+  private S3Path(FileSystem fileSystem, String bucket, String key) {
+    // Bucket name rules:
+    // http://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html
+    // - Bucket name length must be [3~63] characters.
+    // - "Labels" are defined as bucket name parts separated by single period delimiter.
+    // - Each label starts and ends with lowercase letter or number.
+    // - Bucket name cannot be formatted as an IP address.
+    if (bucket == null) {
```

Comment:
Is there a library supplied by Amazon that does this validation for us?

Reply:
I'm not sure if it's in the SDK or an error returned by the AWS API, but either way, it's unnecessary and I reduced this to null-and-empty checks.